Handle all loading states within the `PhoneNumberCard` component

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

It's a follow-up for https://github.com/woocommerce/google-listings-and-ads/pull/917#discussion_r681624815.

- Handle all loading states within the `PhoneNumberCard` component to show more card content as soon as possible.

❗  **This's a refactor PR and didn't affect the functionality of the Contact Information feature. It could be reviewed and (maybe) merged after the Contact Information is released.**

### Screenshots:

#### MC setup

- When the phone number is empty
    ![Kapture 2021-08-05 at 12 19 19](https://user-images.githubusercontent.com/17420811/128292736-97181dff-c072-4505-8be8-caf9cda3f6f3.gif)
- When the phone number exists
    ![Kapture 2021-08-05 at 12 21 30](https://user-images.githubusercontent.com/17420811/128292716-8d0b87ab-dc3f-47f2-ac86-a3c597f8ffd9.gif)

#### Settings

- When refreshing the edit contact info page
    ![Kapture 2021-08-05 at 12 23 15](https://user-images.githubusercontent.com/17420811/128292694-d5cd29bd-121f-4dec-b0ae-ff6bc5d5954b.gif)

### Detailed test instructions:

To enter the MC setup step 4 directly, it could replace [this line in js/src/setup-mc/setup-stepper/saved-setup-stepper.js](https://github.com/woocommerce/google-listings-and-ads/blob/b377bf8e85fceb4bb48e0537150e65c62e97054e/js/src/setup-mc/setup-stepper/saved-setup-stepper.js#L21) by 

```js
const [ step, setStep ] = useState( '4' );
```

To mock the API response of phone number and add more delay to view the loading state, it could add PHP snippet in **google-listings-and-ads.php**.

```php
add_filter(
	'woocommerce_gla_prepared_response_mc-contact-information',
	function( $response ) {
		sleep( 1 );
		return [
			'is_mc_address_different' => false,
			'phone_number' => '+12133734253',
			'wc_address' => [
				'street_address' => '',
			],
		];
	},
);
```

1. Go to the MC setup and settings pages with empty/existed phone number
2. Check the loading spinner is shown within the `PhoneNumberCard` component instead of the parent layer for all cases

### Changelog entry

eason9487 committed:
Handle loading state within PhoneNumberCard

Address https://github.com/woocommerce/google-listings-and-ads/pull/917#discussion_r681624815

eason9487 committed:
Add JSDoc for PhoneNumberCard

ecgan commented:
@eason9487 , for the settings page: 

![](https://user-images.githubusercontent.com/17420811/128292694-d5cd29bd-121f-4dec-b0ae-ff6bc5d5954b.gif)

Do we want to show the "horizontal divider" line when the loading is in progress? I think it would look nicer, and it would be consistent with the store address UI below it too.

ecgan commented on js/src/components/contact-information/phone-number-card.js:
```diff
@@ -105,67 +128,114 @@ function PhoneNumberContent( {
+	);
+}
+
+/**
+ * Renders phone number data in Card UI and is able to edit.
+ *
+ * @param {Object} props React props.
+ * @param {PhoneNumber} props.phoneNumber Phone number data.
+ * @param {boolean} [props.isPreview=false] Whether to display as preview UI.
+ * @param {boolean|null} [props.initEditing=null] Specify the inital UI state. This prop would be ignored if `isPreview` is true.
+ *     `true`: initialize with the editing UI.
+ *     `false`: initialize with the viewing UI.
+ *     `null`: determine the initial UI state according to the `data.isValid` after the `phoneNumber` loaded.
+ * @param {Function} [props.onEditClick] Called when clicking on "Edit" button.
+ *     If this callback is omitted, it will enter edit mode when clicking on "Edit" button.
+ * @param {onPhoneNumberChange} [props.onPhoneNumberChange] Called when inputs of phone number are changed in edit mode.
+ */
```
Thank you for the JSDoc! 😻

ecgan commented on js/src/components/contact-information/phone-number-card.js:
```diff
@@ -105,67 +128,114 @@ function PhoneNumberContent( {
 
-	let description = null;
-	let phoneNumberContent = null;
+	useEffect( () => {
+		if ( loaded && isEditing === null ) {
+			setEditing( ! data.isValid );
+		}
+	}, [ loaded, data.isValid, isEditing ] );
```
💅  It took me a while to figure out this relates to your JSDoc comment `` "determine the initial UI state according to the `data.isValid` after the `phoneNumber` loaded." ``. Would be good to have a comment to explain this, or even refactor it.

eason9487 replied:
Added a comment by https://github.com/woocommerce/google-listings-and-ads/pull/934/commits/63460f79b5c69866d7559ce8af57e36e26681fbd.

ecgan commented on js/src/components/contact-information/phone-number-card.js:
```diff
@@ -105,67 +128,114 @@ function PhoneNumberContent( {
-			phoneNumberContent = <AppSpinner />;
-		}
+	let description = null;
+	let indicator = null;
```
💅  I'd specify `Spinner` as the default indicator so that we don't need the `else` block below.

```suggestion
	let indicator = <Spinner />;
```

ecgan reviewed (approved):
Tested, working as expected. Left a few 💅  comments. Approving. 👍

eason9487 committed:
Show the divider when loading

Address https://github.com/woocommerce/google-listings-and-ads/pull/934#issuecomment-893476476

eason9487 committed:
Add a comment for the handling of initEditing = null

Address https://github.com/woocommerce/google-listings-and-ads/pull/934#discussion_r683568574

eason9487 commented:
@ecgan

> Do we want to show the "horizontal divider" line when the loading is in progress? I think it would look nicer, and it would be consistent with the store address UI below it too.

Thank you for noticing this. I added by https://github.com/woocommerce/google-listings-and-ads/pull/934/commits/3897636ee8449b124f3d6a21ca01d2b9543c135e and it looks nicer. 👍 

![Kapture 2021-08-06 at 10 22 18](https://user-images.githubusercontent.com/17420811/128448436-85e600c0-c4f3-4e29-b66d-285bca66d455.gif)

eason9487 committed:
Simplify the if else statements

Address https://github.com/woocommerce/google-listings-and-ads/pull/934#discussion_r683571373

Co-authored-by: Gan Eng Chin <ecgan@users.noreply.github.com>
